File: python/open_guardrail/guards/version_info_detect.py
"""Detect software version information leaks."""
from __future__ import annotations

import re
import time
from typing import Optional

from open_guardrail.core import GuardResult

_DEFAULT_PATTERNS: list[re.Pattern[str]] = [
    re.compile(r"Apache/\d+\.\d+"),
    re.compile(r"nginx/\d+\.\d+"),
    re.compile(r"PHP/\d+\.\d+"),
    re.compile(r"Node\.js/v?\d+\.\d+"),
    re.compile(r"Python/\d+\.\d+"),
    re.compile(
        r"Server:\s*\w+/\d+\.\d+", re.IGNORECASE
    ),
    re.compile(
        r"X-Powered-By:\s*\w+", re.IGNORECASE
    ),
    re.compile(r"Microsoft-IIS/\d+\.\d+"),
    re.compile(r"OpenSSL/\d+\.\d+"),
    re.compile(r"Express/\d+\.\d+"),
    re.compile(r"Django/\d+\.\d+"),
    re.compile(r"Rails/\d+\.\d+"),
]


class _VersionInfoDetect:
    def __init__(self, *, action: str = "block") -> None:
        self.name = "version-info-detect"
        self.action = action
        self._patterns = list(_DEFAULT_PATTERNS)

    def check(
        self, text: str, stage: str = "output"
    ) -> GuardResult:
        start = time.perf_counter()
        matched: list[str] = []

        for pat in self._patterns:
            if pat.search(text):
                matched.append(pat.pattern)

        triggered = len(matched) > 0
        score = (
            min(len(matched) / 3, 1.0)
            if triggered
            else 0.0
        )
        elapsed = (time.perf_counter() - start) * 1000

        return GuardResult(
            guard_name="version-info-detect",
            passed=not triggered,
            action=self.action if triggered else "allow",
            score=score,
            message=(
                "Software version info detected"
                if triggered
                else None
            ),
            latency_ms=round(elapsed, 2),
            details=(
                {"matched_patterns": len(matched)}
                if triggered
                else None
            ),
        )


def version_info_detect(
    *, action: str = "block"
) -> _VersionInfoDetect:
    return _VersionInfoDetect(action=action)

File: packages/guards/src/version-info-detect.ts
import type { Guard, GuardResult, GuardContext } from 'open-guardrail-core';

interface VersionInfoDetectOptions {
  action: 'block' | 'warn';
}

const PATTERNS: RegExp[] = [
  /Apache\/\d+\.\d+/,
  /nginx\/\d+\.\d+/,
  /PHP\/\d+\.\d+/,
  /Node\.js\/v?\d+\.\d+/,
  /Python\/\d+\.\d+/,
  /Server:\s*\w+\/\d+\.\d+/i,
  /X-Powered-By:\s*\w+/i,
  /Microsoft-IIS\/\d+\.\d+/,
  /OpenSSL\/\d+\.\d+/,
  /Express\/\d+\.\d+/,
  /Django\/\d+\.\d+/,
  /Rails\/\d+\.\d+/,
];

export function versionInfoDetect(options: VersionInfoDetectOptions): Guard {
  return {
    name: 'version-info-detect',
    version: '0.1.0',
    description: 'Detects software version information leaks',
    category: 'security',
    supportedStages: ['output'],
    async check(text: string, _ctx: GuardContext): Promise<GuardResult> {
      const start = performance.now();
      const matched: string[] = [];

      for (const pattern of PATTERNS) {
        if (pattern.test(text)) {
          matched.push(pattern.source);
        }
      }

      const triggered = matched.length > 0;
      const score = triggered ? Math.min(matched.length / 3, 1.0) : 0;

      return {
        guardName: 'version-info-detect',
        passed: !triggered,
        action: triggered ? options.action : 'allow',
        score,
        latencyMs: Math.round(performance.now() - start),
        details: triggered ? { matchedPatterns: matched.length } : undefined,
      };
    },
  };
}
